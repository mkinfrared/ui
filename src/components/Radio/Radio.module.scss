@import "../../styles/common";

.Radio {
  --transition-time: 150ms;

  display: flex;
  width: fit-content;

  &.error {
    color: var(--error);

    .circle {
      &::before,
      &::after {
        background: var(--error-gradient);
      }
    }
  }

  .container {
    align-items: center;
    border-radius: 0.4rem;
    display: flex;
  }

  .input {
    height: 0;
    margin: 0;
    opacity: 0;
    position: absolute;
    width: 0;
    z-index: -1;
  }

  .input:checked ~ .container .circle {
    &::after {
      opacity: 1;

      /**
        google chrome has an issue rendering before and after
        elements when height is a float number
      */
      // transform: scale(0.6);
    }
  }

  .input:focus-visible ~ .container {
    @include focus-visible;
  }

  &.readOnly {
    pointer-events: none;
  }

  &.readOnly .input,
  .input:disabled {
    ~ .container {
      @include disabled;

      .text {
        color: var(--grey);
      }
    }
  }

  .circle {
    aspect-ratio: 1/1;
    height: var(--text-height);
    position: relative;

    &::before,
    &::after {
      background-image: var(--primary-gradient);
      border-radius: 50%;
      content: "";
      inset: 0;
      padding: 0.3rem;
      position: absolute;
    }

    &::before {
      border-radius: 50%;
      /* stylelint-disable */
      mask: linear-gradient(#ffffff 0 0) content-box,
        linear-gradient(#ffffff 0 0);
      /* stylelint-enable */
      mask-composite: exclude;
      transition: all 1ms ease-in-out;
      transition-delay: 0s;
      z-index: -1;
    }

    &::after {
      border-radius: 50%;
      opacity: 0;

      /**
        google chrome has an issue rendering before and after
        elements when height is a float number
      */
      transform: scale(0.6);
      transition: all var(--transition-time) ease-in-out;
      z-index: -1;
    }
  }

  .text {
    margin-left: 1rem;
  }
}
