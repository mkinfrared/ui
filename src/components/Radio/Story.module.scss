.Story {
  .radioGroup {
    margin-bottom: 1rem;
  }
}
