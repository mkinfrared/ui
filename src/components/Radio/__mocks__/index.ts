export { default } from "./Radio.mock";
