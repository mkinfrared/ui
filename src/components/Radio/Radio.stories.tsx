import { Meta, Story } from "@storybook/react";
import React, { useState } from "react";

import { Radio } from "./Radio";
import { RadioProps } from "./Radio.type";
import css from "./Story.module.scss";

export default {
  title: "UI/Radio",
  component: Radio,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
} as Meta;

const UncontrolledTemplate: Story<RadioProps> = (args) => (
  <div className={css.Story}>
    <Radio {...args} />
  </div>
);

const ControlledTemplate: Story<RadioProps> = () => {
  const [checked, setChecked] = useState(false);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { target } = event;

    setChecked(target.checked);
  };

  return (
    <div className={css.Story}>
      <Radio checked={checked} onChange={handleChange} />
    </div>
  );
};

const MultipleUncontrolledTemplate: Story<RadioProps> = () => (
  <div className={css.Story}>
    <form>
      <div className={css.radioGroup}>
        <Radio name="color" label="red" />
        <Radio name="color" label="green" />
        <Radio name="color" label="blue" />
        <Radio name="color" label="rainbow" />
      </div>
      <div className={css.radioGroup}>
        <Radio name="name" label="Timmy" />
        <Radio name="name" label="Jimmy" />
        <Radio name="name" label="Kenny" />
        <Radio name="name" label="Randy" />
      </div>
    </form>
  </div>
);

const MultipleControlledTemplate: Story<RadioProps> = () => {
  const [value, setValue] = useState<string | undefined>();

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { target } = event;

    setValue(target.value);
  };

  return (
    <div className={css.Story}>
      <Radio
        checked={value === "red"}
        name="color"
        onChange={handleChange}
        label="red"
        value="red"
      />
      <Radio
        checked={value === "green"}
        name="color"
        onChange={handleChange}
        label="green"
        value="green"
      />
      <Radio
        checked={value === "blue"}
        name="color"
        onChange={handleChange}
        label="blue"
        value="blue"
      />
      <Radio
        checked={value === "rainbow"}
        name="color"
        onChange={handleChange}
        label="rainbow"
        value="rainbow"
      />
    </div>
  );
};

const Uncontrolled = UncontrolledTemplate.bind({});
const Controlled = ControlledTemplate.bind({});
const MultipleUncontrolled = MultipleUncontrolledTemplate.bind({});
const MultipleControlled = MultipleControlledTemplate.bind({});

Uncontrolled.args = {
  label: "marklar",
};

Uncontrolled.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Uncontrolled, Controlled, MultipleUncontrolled, MultipleControlled };
