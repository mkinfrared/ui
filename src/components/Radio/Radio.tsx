import { memo } from "react";

import Text from "components/Text";
import { classNames, mergeRefs } from "utils";

import css from "./Radio.module.scss";
import { RadioProps } from "./Radio.type";

const Radio = ({
  checked,
  className,
  defaultChecked,
  disabled = false,
  error,
  inputRef,
  label,
  name,
  onChange,
  readOnly = false,
  value,
}: RadioProps) => {
  const refs = mergeRefs(inputRef);

  return (
    <label
      className={classNames(
        css.Radio,
        error && css.error,
        readOnly && css.readOnly,
        className,
      )}
      data-testid="Radio"
    >
      <input
        className={css.input}
        disabled={disabled}
        defaultChecked={defaultChecked}
        ref={refs}
        name={name}
        checked={checked}
        onChange={onChange}
        readOnly={readOnly}
        type="radio"
        value={value}
      />
      <div className={css.container}>
        <div className={css.circle} />
        {label && <Text className={css.text}>{label}</Text>}
      </div>
    </label>
  );
};

export { Radio };

export default memo(Radio);
