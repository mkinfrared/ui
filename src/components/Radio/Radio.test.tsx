import { render } from "@testing-library/react";
import React from "react";

import { Radio } from "./Radio";

describe("<Radio />", () => {
  const Component = <Radio />;

  it("should be defined", () => {
    expect(Radio).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should contain a data test id", () => {
    const { getByTestId } = render(Component);
    const element = getByTestId("Radio");

    expect(element).toBeDefined();
  });

  it("should render <Text /> when label is truthy", () => {
    const label = "markar";
    const { getByTestId } = render(<Radio label={label} />);
    const text = getByTestId("Text");

    expect(text).toBeDefined();
  });

  it("should add an 'error' class to label when error is true", () => {
    const { getByTestId } = render(<Radio error />);
    const label = getByTestId("Radio");

    expect(label.classList).toContain("error");
  });
});
