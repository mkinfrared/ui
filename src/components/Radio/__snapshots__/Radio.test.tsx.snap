// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should match the snapshot 1`] = `
<div>
  <label
    class="Radio"
    data-testid="Radio"
  >
    <input
      class="input"
      type="radio"
      value=""
    />
    <div
      class="container"
    >
      <div
        class="circle"
      />
    </div>
  </label>
</div>
`;
